runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -10258,41 +10253,46 @@ genHeapAlloc(TR::Node * node, TR::Instruction *& iCursor, bool isVariableLen, TR
          iCursor = generateRXInstruction(cg, TR::InstOpCode::getStoreOpCode(), node, sizeReg,
                       generateS390MemoryReference(metaReg, offsetof(J9VMThread, heapAlloc), cg), iCursor);
       TR::LabelSymbol * fillerRemLabel = generateLabelSymbol(cg);
-      TR::LabelSymbol * doneLabel = generateLabelSymbol(cg);
 
       TR::LabelSymbol * fillerLoopLabel = generateLabelSymbol(cg);
 
       // do this clear, if disableBatchClear is on
-      if (disableBatchClear && disableInitClear==NULL) //&& (node->getOpCodeValue() == TR::anewarray) && (node->getFirstChild()->getInt()>0) && (node->getFirstChild()->getInt()<6) )
+      if (clearMemory)
          {
          iCursor = generateRRInstruction(cg, TR::InstOpCode::getLoadRegOpCode(), node, addressReg, resReg, iCursor);
-         // Dont overwrite the class
-         //
-         iCursor = generateRRInstruction(cg, TR::InstOpCode::getLoadRegOpCode(), node, shiftReg, lengthReg, iCursor);
-         iCursor = generateRSInstruction(cg, TR::InstOpCode::SRA, node, shiftReg, 8);
 
+         if (TR::comp()->target().is64Bit())
+            {
+            iCursor = generateRSInstruction(cg, TR::InstOpCode::SRAG, node, shiftReg, lengthReg, 8, iCursor);
+            }
+         else
+            {
+            if (comp->target().cpu.isAtLeast(OMR_PROCESSOR_S390_Z196))
+               {
+               iCursor = generateRSInstruction(cg, TR::InstOpCode::SRAK, node, shiftReg, lengthReg, 8, iCursor);
```

Comment:
Not sure if `SRAK` is safe here since left most bit is the sign and not touched by shift operation!
It is unlikely to have that large object on 32bit configuration but it is not safe to make assumptions.

Reply:
OMR uses the same logic for [arraycmp](https://github.com/eclipse/omr/blob/c2719521f1311f3b38c4ceddfc4c7adcbf8cbc80/compiler/z/codegen/OMRTreeEvaluator.cpp#L10964) so maybe it is fine!